lib/floe/workflow.rb

```diff
@@ -46,48 +29,63 @@ def initialize(payload, context = nil, credentials = {})
       @states         = payload["States"].to_a.map { |name, state| State.build!(self, name, state) }
       @states_by_name = @states.each_with_object({}) { |state, result| result[state.name] = state }
 
-      context.state["Name"] ||= start_at
+      unless context.state.key?("Name")
+        context.state["Name"] = start_at
+        context.state["Input"]     = context.execution["Input"].dup
+      end
     rescue JSON::ParserError => err
       raise Floe::InvalidWorkflowError, err.message
     end
 
-    def run!
-      step until end?
+    def run!(timeout: :forever)
+
+      while !end? && step(:timeout => timeout) == 0 
+      end
```

Comment:
Would `loop while` work here?

```suggestion
      loop while !end? && step(:timeout => timeout) == 0 
```